parameters/src/lib.rs

```diff
@@ -0,0 +1,105 @@
+//! # Parameters
+//! Offer a centra place to store and configure parameters.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::unused_unit)]
+
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+
+use frame_support::traits::EnsureOriginWithArg;
+use orml_traits::parameters::{AggregratedKeyValue, Key, ParameterStore};
+
+mod mock;
+mod tests;
+mod weights;
+
+pub use module::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The key value type for parameters. Usually created by
+		/// orml_traits::parameters::define_aggregrated_parameters
+		type AggregratedKeyValue: AggregratedKeyValue;
+
+		/// The origin which may update the parameter.
+		type AdminOrigin: EnsureOriginWithArg<Self::RuntimeOrigin, KeyOf<Self>>;
+
+		/// Weight information for extrinsics in this module.
+		type WeightInfo: WeightInfo;
+	}
+
+	type KeyOf<T> = <<T as Config>::AggregratedKeyValue as AggregratedKeyValue>::AggregratedKey;
+	type ValueOf<T> = <<T as Config>::AggregratedKeyValue as AggregratedKeyValue>::AggregratedValue;
+
+	#[pallet::error]
+	pub enum Error<T> {}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Parameter is updated
+		Updated { key_value: T::AggregratedKeyValue },
+	}
+
+	/// Stored parameters.
+	///
+	/// map KeyOf<T> => Option<ValueOf<T>>
+	#[pallet::storage]
+	pub type Parameters<T: Config> = StorageMap<_, Blake2_128Concat, KeyOf<T>, ValueOf<T>, OptionQuery>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Set parameter
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::set_parameter())]
+		pub fn set_parameter(origin: OriginFor<T>, key_value: T::AggregratedKeyValue) -> DispatchResult {
+			let (key, value) = key_value.clone().into_parts();
+
+			T::AdminOrigin::ensure_origin(origin, &key)?;
+
+			Parameters::<T>::mutate(key, |v| *v = value);
+
+			Self::deposit_event(Event::Updated { key_value });
```

Comment:
It would be better to first `ensure_origin` before doing any further work.

```suggestion
			T::AdminOrigin::ensure_origin(origin, &key)?;
			
			let (key, value) = key_value.clone().into_parts();
			Parameters::<T>::mutate(key, |v| *v = value);

			Self::deposit_event(Event::Updated { key_value });
```

Reply:
can't do this, the ensure origin check requires the key

Reply:
ah, sorry, missed that 👍